### PR TITLE
Fix remark plugin for diagram/mermaidjs not working

### PR DESCRIPTION
**Pull request by YoniFeng**
Using the remark plugin as documented (to enable mermaid diagrams in md/mdx) fails:
`ENOENT: no such file or directory, open 'C:\projects\YoniFeng-blog\node_modules\mermaid\dist\mermaid.js'`

The issue is that mermaid switched to ESM modules. `mermaid.js` no longer exists.
I tried naively using the ESM module (and adding `type: "module"` to the puppeteer's launch args), but that didn't work.

Downgrading dependency as a quick fix.

### CODE REVIEW

**Comment by YoniFeng** [comment]
btw I wasn't able to re-build `dist/index.js`.

```
❯ pnpm run build

> astro-diagram@0.7.0 build C:\oss\web-garden\components\Diagram
> parcel build index.ts

🚨 Build failed.

@parcel/core: No transformers found for Diagram.astro.

  C:\OSS\web-garden\components\Diagram\node_modules\.pnpm\@parcel+config-default@2.7.0_@parcel+core@2.7.0\node_modules\@parcel\config-default\index.json:3:3
     2 |   "bundler": "@parcel/bundler-default",
  >  3 |   "transformers": {
  >    |   ^^^^^^^^^^^^^^^^^
  >  4 |     "types:*.{ts,tsx}": ["@parcel/transformer-typ
  >    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  >  5 |     "bundle-text:*": ["...", "@parcel/transformer
  >    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  >  6 |     "data-url:*": ["...", "@parcel/transformer-in
  >    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  >  7 |     "worklet:*.{js,mjs,jsm,jsx,es6,cjs,ts,tsx}":
  >    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  >  8 |       "@parcel/transformer-worklet",
  >    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  >  9 |       "..."
  >    | ^^^^^^^^^^^
  > 10 |     ],
  >    | ^^^^^^
  > 11 |     "*.{js,mjs,jsm,jsx,es6,cjs,ts,tsx}": [
  >    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  > 12 |       "@parcel/transformer-babel",
  >    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  > 13 |       "@parcel/transformer-js",
  >    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Am I doing something wrong?

**Comment by jackcarey** [comment]
Thank you, this is a much needed fix for my sites too

**Comment by JulianCataldo** [comment]
Thank you for chiming in!

I admit I don't wan't to maintain library relying on `puppeteer` anymore (or any big binaries stuff really, even `sharp` has been annoying twice for me).

Like OG Image generation with Satori, I believe we must find a way to not spin up a whole browser for generating mermaid diagram, server-side.

When I built this, I was conscious of those limitations, but I wanted to see how far we could get.
In practice, it's cumbersome to put puppeteer anywhere in your pipeline. Even with all caching tricks you can find.

I think @remcohaszing has been doing some work here, but I didn't dug that yet!

- https://github.com/remcohaszing/mermaid-isomorphic
- https://github.com/remcohaszing/remark-mermaidjs

---

> btw I wasn't able to re-build `dist/index.js`.

I was not really accustomed to building / packaging at that time, so I used parcel, but really, just `tsc` could have done the job here.

Oh, and good news, you don't need to transpile to JS anymore in order to load TypeScript stuff from the Astro config file (with MJS). I found that by accident.

So basically you could import the source directly.

---

I keep this open because pinning a older version cannot be considered as a "fix".

Cheers

**Comment by YoniFeng** [comment]
Both projects you linked use either Pupeteer or Playwright.
I understand this is far from an ideal solution and probably becomes a build bottleneck for sufficiently large projects (that make extensive use of mermaid diagrams), but implementing an ideal solution probably isn't trivial and would likely be a considerable maintenance burden.

I agree this doesn't fix the issue, it's just a quick hack I did to unblock myself.
Closing this PR.

**Comment by remcohaszing** [comment]
You need to use a browser. See https://github.com/remcohaszing/remark-mermaidjs/issues/3 for context.

You can't load ESM in Chrome / WebKit using a `file:` protocol. See https://github.com/microsoft/playwright/issues/22292

Mermaid moved to ESM only in version 10. The UMD bundle is being added back in https://github.com/mermaid-js/mermaid/pull/4281. For now, you can use a pinned version of `10.0.3-alpha.1`.

As @JulianCataldo pointed out, I recently created [`mermaid-isomorphic`](https://github.com/remcohaszing/mermaid-isomorphic) to take away most of the pain of working with Mermaid in Node.js. I haven't released it yet though.
